### ISSUE TITLE
Erroneous "use of infinity" warning

### ISSUE THREAD

**Issue by ahatanak** [llvm/llvm-project#123231]
clang incorrectly emits a warning when a method called infinity is called.

$ cat test.cpp
```
double infinity() { return 0; }

int main() {
   return infinity();
}
```

$ clang++ -ffast-math test.cpp -c
test.cpp:4:11: warning: use of infinity is undefined behavior due to the currently enabled floating-point options [-Wnan-infinity-disabled]
    4 |    return infinity();

**Comment by ahatanak** [comment]
The warning was introduced in 1e2a4ccb6237b6db9db7dc3fd11a94410ec9b0df. @zahiraam

**Comment by shafik** [comment]
A godbolt sample is always helpful: https://godbolt.org/z/WrYb86qqq

**Comment by shafik** [comment]
Yeah, it looks like the diagnostic needs to be tightened up a bit.

CC @AaronBallman

**Comment by erichkeane** [comment]
Hmm... this looks like something where `CheckInfNaNFunction` in the original patch should be checking the 'builtin', not the name of the function? This part: `FDecl->getIdentifier() && FDecl->getName() == "infinity")` is misguided perhaps.

That is checking ANY function named infinity for the purposes of this warning, which isn't correct.

**Comment by AaronBallman** [comment]
> That is checking ANY function named infinity for the purposes of this warning, which isn't correct.

Yeah, that seems like it should be looking for `std::numeric_limits<T>::infinity()` ?

**Comment by AmrDeveloper** [comment]
> > That is checking ANY function named infinity for the purposes of this warning, which isn't correct.
> 
> Yeah, that seems like it should be looking for `std::numeric_limits<T>::infinity()` ?

I think we should add checks for the `MethodDecl` parent name and namespace

I am interested to work on it :D

**Comment by erichkeane** [comment]
> > > That is checking ANY function named infinity for the purposes of this warning, which isn't correct.
> > 
> > 
> > Yeah, that seems like it should be looking for `std::numeric_limits<T>::infinity()` ?
> 
> I think we should add checks for the `MethodDecl` parent name and namespace
> 
> I am interested to work on it :D

yeah, something like that?  I don't know the original intent of this, but I imagine it means numeric_limits::infinity?  The original author is unfortunately the only one who knows as far as I can tell, as this wasn't mentioned/discussed in the original patch.

As far as checks; Yes, you have to check the `CXXRecordDecl` of it, as well as that one's namespace.  We should have prior art.

Feel free to start working on it!

**Comment by AmrDeveloper** [comment]
> I imagine it means numeric_limits::infinity?

From the test files, I think yes because there are cases for `numeric_limits::infinity` but no case with normal function that has name `infinity`

**Comment by zahiraam** [comment]
Will have a fix soon in a couple of weeks if that can wait.

**Comment by erichkeane** [comment]
> > I imagine it means numeric_limits::infinity?
> 
> From the test files, I think yes because there are cases for `numeric_limits::infinity` but no case with normal function that has name `infinity`

Cool!  Then we should probably check for `CXXMethodDecl`, name, `CXXRecordDecl` + name, and NS.  Looks like @zahiraam can't get to it for a while, so have at it @AmrDeveloper

**Comment by zahiraam** [comment]
Thanks!
